src/main/java/com/nha/abdm/wrapper/hrp/hipInitiatedLinking/serviceImpl/HipInitiatedServiceImpl.java

```diff
@@ -0,0 +1,151 @@
+    @Autowired
+    PatientTableService patientTableService;
+
+    @Autowired
+    PatientRepo patientRepo;
+    @Autowired
+    SessionManager sessionManager;
+
+    @Autowired
+    LogsRepo logsRepo;
+    @Autowired
+    CustomError customError;
+    private WebClient.Builder webClientBuilder = WebClient.builder();
+    private ResponseEntity<ObjectNode> responseEntity;
+
+    public Object authInit(LinkRecordsResponse data) throws URISyntaxException, JsonProcessingException {
+        try{
+            Patients existingRecord=patientRepo.findByPatientReference(data.getPatientReference());
+            if(existingRecord==null || existingRecord.getAbhaAddress()==null){
+                customError.setCode(10001);
+                customError.setMessage("Abha address not found with patient UID , kindly add patient");
+                log.info("Patient not found");
+                logsTableService.setContent(data,new HttpEntity<>(new ObjectNode(JsonNodeFactory.instance).put("Error","Patient not found, kindly add patient")), Object.class);
+                return customError;
+            }
+            patientTableService.addPatient(data);
+            HttpEntity<ObjectNode> requestEntity = AuthInitBody.builder()
+                    .data(data)
+                    .abhaAddress(existingRecord.getAbhaAddress())
+                    .sessionManager(sessionManager)
+                    .build()
+                    .makeRequest();
+            log.info("LinkRecords storing data");
+            logsTableService.setContent(data,requestEntity, LinkRecordsResponse.class);
+
+            responseEntity=MakeRequest.post(GatewayApiPaths.LINK_AUTH_INIT,requestEntity);
+
+            log.info(responseEntity.getStatusCode());
+            ObjectNode responseData = new ObjectNode(JsonNodeFactory.instance);
+            responseData.put("Status Code : ", responseEntity.getStatusCode().toString());
+            responseData.put("requestId :", ((ObjectNode)requestEntity.getBody()).get("requestId"));
+            return responseData;
+
+        }catch(Exception e){
+            log.info("Link authInit : "+e);
+            customError.setCode(1000);
+            customError.setMessage("Error Linking careContexts");
+            return customError;
+        }
+    }
+    public ResponseEntity<ObjectNode> confirmAuth(OnInitResponse data) throws URISyntaxException, JsonProcessingException, FileNotFoundException, TimeoutException {
+        RequestLogs existingRecord=logsRepo.findByGatewayRequestId1(data.getResp().getRequestId());
+        if(existingRecord!=null){
+            log.info("In confirmAuth found existing record");
+            String otp=null;
+            if(data.getAuth().getMode().equals("MOBILE_OTP")){
+                log.info("confirmAuth : "+"Waiting for otp......");
+                long startTime = System.currentTimeMillis();
+                long timeout = TimeUnit.SECONDS.toMillis(120);
+                while (otp == null) {
```

Comment:
This logic will hold the `gateway` request to be on hold until the OTP is received from the User into the Wrapper. Better way could be to acknowledge the Gateway immediately and wait for the OTP from the user. And when the OTP is received from the User in a different request, we call the Gateway with the received OTP.